File: Initializers.jl
module Initializers

export
initialize,
Initializer,
HeUniform,
HeNormal,
RandomNormal,
RandomUniform,
Zeros,
Ones

include("Math.jl")

using Random

using .Math

abstract type Initializer end

function initialize(initializer::T where T<:Initializer); end

mutable struct HeNormal <: Initializer
    type::Type{T} where T<:Number
    dims::Tuple{Vararg{Integer, N} where N}

    HeNormal(type::Type{T} where T<:Number,
        dims::Tuple{Vararg{Integer, N} where N}) = new(type, dims)
end

initialize(initializer::HeNormal, rng::AbstractRNG = Random.GLOBAL_RNG) =
    Math.he_normal(
        initializer.type,
        initializer.dims,
        size(initializer.dims, 1),
        rng=rng)

mutable struct HeUniform <: Initializer
    type::Type{T} where T<:Number
    dims::Tuple{Vararg{Integer, N} where N}

    HeUniform(type::Type{T} where T<:Number,
        dims::Tuple{Vararg{Integer, N} where N}) = new(type, dims)
end

initialize(initializer::HeUniform, rng::AbstractRNG = Random.GLOBAL_RNG) =
    Math.he_uniform(
        initializer.type,
        initializer.dims,
        size(initializer.dims, 1),
        rng=rng)

mutable struct RandomUniform <: Initializer
    type::Type{T} where T<:Number
    dims::Tuple{Vararg{Integer, N} where N}

    RandomUniform(type::Type{T} where T<:Number,
        dims::Tuple{Vararg{Integer, N} where N}) = new(type, dims)
end

initialize(initializer::RandomUniform, rng::AbstractRNG = Random.GLOBAL_RNG) =
    Math.random_uniform(
        initializer.type,
        initializer.dims,
        rng=rng)

mutable struct RandomNormal <: Initializer
    type::Type{T} where T<:Number
    dims::Tuple{Vararg{Integer, N} where N}

    RandomNormal(type::Type{T} where T<:Number,
        dims::Tuple{Vararg{Integer, N} where N}) = new(type, dims)
end

initialize(initializer::RandomNormal, rng::AbstractRNG = Random.GLOBAL_RNG) =
    Math.random_normal(
        initializer.type,
        initializer.dims,
        rng=rng)

mutable struct Zeros <: Initializer
    type::Type{T} where T<:Number
    dims::Tuple{Vararg{Integer, N} where N}

    Zeros(type::Type{T} where T<:Number,
        dims::Tuple{Vararg{Integer, N} where N}) = new(type, dims)
end

initialize(initializer::Zeros, rng::Union{AbstractRNG, Nothing} = nothing) = zeros(
    initializer.type, initializer.dims)

mutable struct Ones <: Initializer
    type::Type{T} where T<:Number
    dims::Tuple{Vararg{Integer, N} where N}

    Ones(type::Type{T} where T<:Number,
        dims::Tuple{Vararg{Integer, N} where N}) = new(type, dims)
end

initialize(initializer::Ones, rng::Union{AbstractRNG, Nothing} = nothing) = ones(
    initializer.type, initializer.dims)

end

File: edunet/edunet/initializers.py
import abc
from typing import Union, Iterable

import numpy as np
from numpy import ndarray
from numpy.random.mtrand import RandomState

import edunet.math as edumath
from edunet.variable import Variable


class Initializer(abc.ABC):
    def __init__(self, shape, dtype):
        self._shape = shape
        self._dtype = dtype

    @abc.abstractmethod
    def initialize(self, random_state: Union[None, int, ndarray, Iterable, float, RandomState] = None) -> Variable:
        pass


class HeNormal(Initializer):
    def initialize(self, random_state=None):
        return Variable(edumath.he_normal(
            self._shape,
            self._shape[0],
            dtype=self._dtype,
            random_state=random_state))


class HeUniform(Initializer):
    def initialize(self, random_state=None):
        return Variable(edumath.he_uniform(
            self._shape,
            self._shape[0],
            dtype=self._dtype,
            random_state=random_state))


class RandomNormal(Initializer):
    def initialize(self, random_state=None):
        return Variable(edumath.random_normal(
            self._shape,
            dtype=self._dtype,
            random_state=random_state))


class RandomUniform(Initializer):
    def initialize(self, random_state=None):
        return Variable(edumath.random_uniform(
            self._shape,
            dtype=self._dtype,
            random_state=random_state))


class Ones(Initializer):
    def initialize(self, random_state=None):
        return Variable(np.ones(self._shape, self._dtype))


class Zeros(Initializer):
    def initialize(self, random_state=None):
        return Variable(np.zeros(self._shape, self._dtype))
